polkadot/xcm/xcm-builder/src/asset_exchange.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Adapters that implement the [`xcm_executor::traits::AssetExchange`] trait.
+
+use core::marker::PhantomData;
+use frame_support::traits::tokens::fungibles;
+use pallet_asset_conversion::{QuoteExchangePrice, SwapCredit};
+use sp_runtime::traits::Zero;
+use sp_std::vec;
+use xcm::prelude::*;
+use xcm_executor::{
+	traits::{AssetExchange, MatchesFungibles},
+	AssetsInHolding,
+};
+
+/// An adapter from [`pallet_asset_conversion::SwapCredit`] and
+/// [`pallet_asset_conversion::QuoteExchangePrice`] to [`xcm_executor::traits::AssetExchange`].
+///
+/// This adapter takes just one fungible asset in `give` and allows only one fungible asset in
+/// `want`. If you need to handle more assets in either `give` or `want`, then you should use
+/// another type that implements [`xcm_executor::traits::AssetExchange`] or build your own.
+pub struct FungiblesPoolAdapter<AssetConversion, Fungibles, Matcher, AccountId>(
+	PhantomData<(AssetConversion, Fungibles, Matcher, AccountId)>,
+);
+impl<AssetConversion, Fungibles, Matcher, AccountId> AssetExchange
+	for FungiblesPoolAdapter<AssetConversion, Fungibles, Matcher, AccountId>
+where
+	AssetConversion: SwapCredit<
+			AccountId,
+			Balance = u128,
+			AssetKind = Fungibles::AssetId,
+			Credit = fungibles::Credit<AccountId, Fungibles>,
+		> + QuoteExchangePrice<Balance = u128, AssetKind = Fungibles::AssetId>,
+	Fungibles: fungibles::Balanced<AccountId, Balance = u128>,
+	Matcher: MatchesFungibles<Fungibles::AssetId, Fungibles::Balance>,
+{
+	fn exchange_asset(
+		_: Option<&Location>,
+		give: AssetsInHolding,
+		want: &Assets,
+		maximal: bool,
+	) -> Result<AssetsInHolding, AssetsInHolding> {
+		let give_asset = give.fungible_assets_iter().next().ok_or_else(|| {
+			log::trace!(
+				target: "xcm::FungiblesPoolAdapter::exchange_asset",
+				"No fungible asset was in `give`.",
+			);
+			give.clone()
+		})?;
+		let want_asset = want.get(0).ok_or_else(|| {
+			log::trace!(
+				target: "xcm::FungiblesPoolAdapter::exchange_asset",
+				"No asset was in `want`.",
+			);
+			give.clone()
+		})?;
```

Comment:
we can actually return `OK` if there is no assets in `want` no?

Reply:
True